challenge_15/action_view.rb

```diff
@@ -39,11 +38,13 @@ def root
   def show_data
     @blogs = Blog.all
 
+    # Note! ERB interpolated HTML is escaped unless marked as HTML safe / raw helper used
+
     erb_response = <<~ERB
       <ul>
       <% @blogs.each do |blog| %>
         <li>
-          <strong>Title: <%= CGI.escape_html(blog.title) %> </strong>, Content: <%= CGI.escape_html(blog.content) %>
+          <strong>Title: <%= blog.title %> </strong>, Content: <%= blog.content %>
```

Comment:
The mechanism here is that [`ActionView::Template::Handlers::ERB::Erubi`](https://github.com/rails/rails/blob/v6.1.3.2/actionview/lib/action_view/template/handlers/erb/erubi.rb) is used as the [ERB implementation](https://github.com/rails/rails/blob/v6.1.3.2/actionview/lib/action_view/template/handlers/erb.rb#L14) inside the [ERB template handler](https://github.com/rails/rails/blob/v6.1.3.2/actionview/lib/action_view/template/handlers.rb#L14). This subclasses [`Erubi::Engine`](https://github.com/jeremyevans/erubi/blob/1.10.0/lib/erubi.rb#L47) from the [`erubi` gem](https://github.com/jeremyevans/erubi), and in particular [supplies an empty `ActionView::OutputBuffer`](https://github.com/rails/rails/blob/v6.1.3.2/actionview/lib/action_view/template/handlers/erb/erubi.rb#L32) for the template to be rendered into. `ActionView::OutputBuffer` [is a subclass](https://github.com/rails/rails/blob/v6.1.3.2/actionview/lib/action_view/buffers.rb#L21) of [`ActiveSupport::SafeBuffer`](https://github.com/rails/rails/blob/v6.1.3.2/activesupport/lib/active_support/core_ext/string/output_safety.rb#L134), which [overrides string concatenation](https://github.com/rails/rails/blob/v6.1.3.2/activesupport/lib/active_support/core_ext/string/output_safety.rb#L186-L189) to [call `CGI.escapeHTML` on non-`html_safe?` strings](https://github.com/rails/rails/blob/v6.1.3.2/activesupport/lib/active_support/core_ext/string/output_safety.rb#L294-L296). So yeah, again, we get a feature so convenient that it’s essentially invisible — HTML escaping “just works” rather than us having to remember to call `CGI.escapeHTML` ourselves — only as a result of a complicated arrangement under the hood.

Reply:
Very cool! Thanks for sharing ❤️